core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -439,6 +441,9 @@ private[spark] class TaskSchedulerImpl(
       }
     }
 
+    val availableSlots = executorIdToCores.values.map(c => c / CPUS_PER_TASK).sum
+    rootPool.updateAvailableSlots(availableSlots)
```

Comment:
this looks a little expensive to be calling in every call to `resourceOffer()` -- most of the time that is called just one executor at a time, and each call does `numExecs` worth of work.  So when you go through scheduling for all executors, you end up doing `O(numExecs^2)` work.      could these two calls be moved into the branch which sets `newExecAvail = true` ?

Reply:
yea, good suggestion.